### PR TITLE
Adopt list commands for editor's "Find References" peek UI

### LINKED ISSUE #21692: Adopt list commands for editor's "Find References" peek UI
Looks like I forgot the adoption in the find references list ([`ReferencesWidget`](https://github.com/Microsoft/vscode/blob/master/src/vs/editor/contrib/referenceSearch/browser/referencesWidget.ts#L484)):

image

//fyi @rebornix

### PR DESCRIPTION
**Pull request by bpasero**
fixes #21692

This adopts the `WorkbenchTree` for the references widget (one of the last that was still using the plain tree). The benefit is that all navigation within the tree, expand/collapse and opening are then configurable via keybindings (`list.*` commands) and consistent with all keybindings in other trees/lists.

**Notable changes:**
* a new command `openReferenceToSide`  is added to specifically handle the case of opening a reference to the side (this requires a new context key `referenceSearchTreeFocused`)
* the controller keyboard handling is removed so that all keyboard access can go through commands from the [workbench](https://github.com/Microsoft/vscode/blob/fix-21692/src/vs/workbench/electron-browser/commands.ts#L28)
* as such, the widget now reacts to `focus` and `selection` events from the tree to open references

### CODE REVIEW

File: src/vs/editor/contrib/referenceSearch/referenceSearch.ts [hunk 1]
```diff
@@ -197,6 +212,16 @@ KeybindingsRegistry.registerCommandAndKeybindingRule({
 	handler: closeActiveReferenceSearch
 });
 
+KeybindingsRegistry.registerCommandAndKeybindingRule({
+	id: 'openReferenceToSide',
+	weight: KeybindingsRegistry.WEIGHT.workbenchContrib(50),
+	primary: KeyMod.CtrlCmd | KeyCode.Enter,
+	mac: {
+		primary: KeyMod.WinCtrl | KeyCode.Enter
```
Comment (jrieken) [inline src/vs/editor/contrib/referenceSearch/referenceSearch.ts]: Why isn't this `CtrlCmd`?

File: src/vs/editor/contrib/referenceSearch/referenceSearch.ts [hunk 2]
```diff
@@ -176,12 +191,12 @@ function closeActiveReferenceSearch(accessor: ServicesAccessor, args: any) {
...
 	}
 
-	controller.closeWidget();
+	fn(controller);
 }
 
 KeybindingsRegistry.registerCommandAndKeybindingRule({
 	id: 'closeReferenceSearch',
-	weight: KeybindingsRegistry.WEIGHT.editorContrib(50),
+	weight: KeybindingsRegistry.WEIGHT.workbenchContrib(50),
```
Comment (jrieken) [inline src/vs/editor/contrib/referenceSearch/referenceSearch.ts]: Why workbench and not editor? This feature is also available in the monaco editor....

File: src/vs/editor/contrib/referenceSearch/referencesWidget.ts [hunk 3]
```diff
@@ -694,32 +638,31 @@ export class ReferenceWidget extends PeekViewWidget {
...
+				keyboardSupport: false
+			};
 
-			this._disposables.push(controller.onDidSelect((element: any) => {
-				if (element instanceof OneReference) {
-					this._onDidSelectReference.fire({ element, kind: 'goto', source: 'tree' });
-				}
-			}));
-			this._disposables.push(controller.onDidOpenToSide((element: any) => {
+			this._tree = this._instantiationService.createInstance(WorkbenchTree, div.getHTMLElement(), config, options);
```
Comment (jrieken) [inline src/vs/editor/contrib/referenceSearch/referencesWidget.ts]: Somewhere the 'cmd+click to open to side' feature is being lost here

**Review by jrieken — CHANGES_REQUESTED** [review]

File: src/vs/editor/contrib/referenceSearch/referenceSearch.ts (hunk 2 above)
Comment (bpasero) [reply src/vs/editor/contrib/referenceSearch/referenceSearch.ts]: @jrieken the workbench is also registering ESC for trees and lists here: https://github.com/Microsoft/vscode/blob/master/src/vs/workbench/electron-browser/commands.ts#L344. This is needed to preserve the behaviour of ESC closing the references widget. It is a bit ugly, I agree. The alternative would be to register this command also from the workbench maybe.

File: src/vs/editor/contrib/referenceSearch/referenceSearch.ts (hunk 1 above)
Comment (bpasero) [reply src/vs/editor/contrib/referenceSearch/referenceSearch.ts]: I tried to be consistent with the other places where we have "Open to Side" (explorer, search, problems):

image

File: src/vs/editor/contrib/referenceSearch/referencesWidget.ts (hunk 3 above)
Comment (bpasero) [reply src/vs/editor/contrib/referenceSearch/referencesWidget.ts]: @jrieken fixed via https://github.com/Microsoft/vscode/pull/41275/commits/64f9e8a98b406d3d04f3934791b98b04d7845719

**Comment by bpasero** [comment]
Also looks like double click was broken for mouse (fixed via https://github.com/Microsoft/vscode/pull/41275/commits/6ebd589d9ad21982cc64ffaaa888f300eb4b8e98). And it looked like each mouse click was triggering the goto definitition twice. Fix is to only react to focus and selection on the tree from keyboard events.

**Comment by bpasero** [comment]
Btw neither in stable nor in insiders does "open to side" using keyboard work for me currently.

**Comment by jrieken** [comment]
> using keyboard work for me currently from the find references tree.

It does for me

jan-09-2018 08-29-22

File: src/vs/editor/contrib/referenceSearch/referencesWidget.ts [hunk 4]
```diff
@@ -695,32 +639,39 @@ export class ReferenceWidget extends PeekViewWidget {
...
+			};
+
+			this._tree = this._instantiationService.createInstance(WorkbenchTree, div.getHTMLElement(), config, options);
+			this._callOnDispose.push(attachListStyler(this._tree, this._themeService));
+
+			ctxReferenceWidgetSearchTreeFocused.bindTo(this._tree.contextKeyService);
+
 			// listen on selection and focus
-			this._disposables.push(controller.onDidFocus((element) => {
+			var onEvent = (element: any, kind: 'show' | 'goto' | 'side') => {
```
Comment (jrieken) [inline src/vs/editor/contrib/referenceSearch/referencesWidget.ts]: `var` 😱

File: src/vs/editor/contrib/referenceSearch/referenceSearch.ts [hunk 5]
```diff
@@ -197,6 +212,16 @@ KeybindingsRegistry.registerCommandAndKeybindingRule({
 	handler: closeActiveReferenceSearch
 });
 
+KeybindingsRegistry.registerCommandAndKeybindingRule({
+	id: 'openReferenceToSide',
+	weight: KeybindingsRegistry.WEIGHT.editorContrib(),
+	primary: KeyMod.CtrlCmd | KeyCode.Enter,
+	mac: {
+		primary: KeyMod.WinCtrl | KeyCode.Enter
```
Comment (jrieken) [inline src/vs/editor/contrib/referenceSearch/referenceSearch.ts]: Why not `cmd` on macs? We use that in the explorer and elsewhere

File: src/vs/editor/contrib/referenceSearch/referencesWidget.ts (hunk 4 above)
Comment (bpasero) [reply src/vs/editor/contrib/referenceSearch/referencesWidget.ts]: @jrieken that is consistent with other usages of `var` in `_fillBody`, I did not want to change that style. There are more `vars` throughout the file, you can clean that up afterwards if you want.

**Comment by bpasero** [comment]
@jrieken mouse works for me, but keyboard does not.

File: src/vs/editor/contrib/referenceSearch/referenceSearch.ts (hunk 5 above)
Comment (bpasero) [reply src/vs/editor/contrib/referenceSearch/referenceSearch.ts]: @jrieken you must have changed the keybindings then, the default is `Ctrl+Enter` on mac for Explorer, Search and Problems view:

image

**Comment by jrieken** [comment]
Yeah, keyboard doesn't work today.

File: src/vs/editor/contrib/referenceSearch/referenceSearch.ts (hunk 5 above)
Comment (jrieken) [reply src/vs/editor/contrib/referenceSearch/referenceSearch.ts]: true story...

**Review by jrieken — APPROVED** [review]

**Pull request merged**